pkg/astnormalization/remove_scalar_duplicates.go

```diff
@@ -0,0 +1,71 @@
+package astnormalization
+
+import (
+	"github.com/jensneuse/graphql-go-tools/pkg/ast"
+	"github.com/jensneuse/graphql-go-tools/pkg/astvisitor"
+)
+
+func removeScalarDuplicates(walker *astvisitor.Walker) {
+	visitor := removeScalarDuplicatesVisitor{
+		Walker: walker,
+	}
+	walker.RegisterScalarTypeDefinitionVisitor(&visitor)
+	walker.RegisterEnterDocumentVisitor(&visitor)
+	walker.RegisterLeaveDocumentVisitor(&visitor)
+	walker.RegisterEnterScalarTypeDefinitionVisitor(&visitor)
+}
+
+type removeScalarDuplicatesVisitor struct {
+	*astvisitor.Walker
+	operation    *ast.Document
+	definition   *ast.Document
+	typesSeen    map[string]bool
+	refsToRemove []int
+	lastRef      int
+}
+
+func (d *removeScalarDuplicatesVisitor) EnterDocument(operation, definition *ast.Document) {
+	d.operation, d.definition = operation, definition
+	d.typesSeen = make(map[string]bool)
+	d.refsToRemove = make([]int, 0)
+	d.lastRef = -1
+}
+
+func (d *removeScalarDuplicatesVisitor) LeaveDocument(operation, definition *ast.Document) {
+	if len(d.refsToRemove) < 1 {
+		return
```

Comment:
There is a function called `DeleteRootNodes` in `pkg/ast/ast.go` file. I think you can use that function to remove root nodes. Just need to collect nodes instead of their references. If it works, that would simplify the code greatly.